File: api-py/app/services/drive_thumb.py
"""Image thumbnail helper for the drive service.

Mirrors api-go/internal/services/drive_thumb.go. Module-level functions so
the DriveService class stays focused on tree CRUD.
"""
from __future__ import annotations

import os
import threading

from .drive import (
    DriveError,
    DriveNotFound,
    DriveNotImage,
    DriveService,
    IMAGE_EXTS,
    THUMB_WIDTH,
)


__all__ = ['make_thumbnail', 'purge_thumb', 'thumb_dir']


_thumb_locks: dict[str, threading.Lock] = {}
_thumb_locks_guard = threading.Lock()


def _key_lock(key: str) -> threading.Lock:
    """Return a per-thumbnail-path lock so two concurrent requests don't
    rebuild the same thumbnail twice.
    """
    with _thumb_locks_guard:
        lock = _thumb_locks.get(key)
        if lock is None:
            lock = threading.Lock()
            _thumb_locks[key] = lock
    return lock


def thumb_dir(base_path: str) -> str:
    return os.path.join(base_path, 'drive', '_thumbs')


def make_thumbnail(service: DriveService, node_id: int) -> str:
    """Build (or reuse) the JPEG thumbnail for an image node and return the
    absolute thumbnail path.

    Raises :class:`DriveNotFound` if the node isn't an active file, or
    :class:`DriveNotImage` if the file isn't a supported image type.
    """
    from PIL import Image, ImageOps

    n = service.find_by_id(node_id)
    if n.type != 'file' or not n.blob_path or n.deleted_at is not None:
        raise DriveNotFound('drive node not found')
    ext = os.path.splitext(n.name)[1].lower()
    if ext not in IMAGE_EXTS:
        raise DriveNotImage('not an image')

    src_abs = service.blob_abs_path(n.blob_path)
    if not os.path.exists(src_abs):
        raise DriveNotFound('source file not found on disk')
    thumbs_dir = thumb_dir(service.base_path)
    os.makedirs(thumbs_dir, exist_ok=True)
    thumb_abs = os.path.join(thumbs_dir, os.path.basename(n.blob_path) + '.jpg')

    if _is_present(thumb_abs):
        return thumb_abs

    with _key_lock(thumb_abs):
        if _is_present(thumb_abs):
            return thumb_abs

        with Image.open(src_abs) as img:
            img = ImageOps.exif_transpose(img)
            w, h = img.size
            if w == 0 or h == 0:
                raise DriveError('zero-dimension image')
            tw = THUMB_WIDTH if w >= THUMB_WIDTH else w
            th = max(1, h * tw // w)
            thumb = img.convert('RGB').resize((tw, th), Image.LANCZOS)

        tmp = thumb_abs + '.part'
        try:
            thumb.save(tmp, format='JPEG', quality=82)
            os.replace(tmp, thumb_abs)
        except Exception:
            try:
                os.remove(tmp)
            except OSError:
                pass
            raise

    return thumb_abs


def purge_thumb(base_path: str, blob_path: str) -> None:
    """Best-effort delete of the cached thumbnail for ``blob_path``."""
    if not blob_path:
        return
    thumb_abs = os.path.join(
        thumb_dir(base_path), os.path.basename(blob_path) + '.jpg'
    )
    try:
        os.remove(thumb_abs)
    except OSError:
        pass


def _is_present(path: str) -> bool:
    return os.path.exists(path) and os.path.getsize(path) > 0

File: api-go/internal/services/drive_thumb.go
package services

import (
	"context"
	"errors"
	"fmt"
	"image/jpeg"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/disintegration/imaging"
)

// thumbWidth is the long-edge in pixels for drive thumbnails.
const thumbWidth = 240

// ErrDriveNotImage is returned by Thumbnail when the requested file isn't an image.
var ErrDriveNotImage = errors.New("not an image")

// imageExts are the extensions we generate thumbnails for. Limited to formats
// `imaging` decodes safely; HEIC etc. are skipped.
var imageExts = map[string]struct{}{
	".jpg": {}, ".jpeg": {}, ".png": {}, ".webp": {}, ".gif": {}, ".bmp": {}, ".tiff": {},
}

// thumbLocks serialises generation of the same thumbnail so concurrent requests
// don't decode/encode the same image twice.
var thumbLocks sync.Map // string -> *sync.Mutex

func keyLock(key string) *sync.Mutex {
	if v, ok := thumbLocks.Load(key); ok {
		return v.(*sync.Mutex)
	}
	mu := &sync.Mutex{}
	actual, _ := thumbLocks.LoadOrStore(key, mu)
	return actual.(*sync.Mutex)
}

// Thumbnail returns the absolute path to a JPEG thumbnail for an image node.
// Generates the thumbnail lazily on first request and caches it next to the
// blob (under drive/_thumbs/). Returns ErrDriveNotImage for non-image nodes.
func (s *DriveService) Thumbnail(ctx context.Context, id int64) (string, error) {
	n, err := s.FindByID(ctx, id)
	if err != nil {
		return "", err
	}
	if n.Type != "file" || !n.BlobPath.Valid || n.DeletedAt.Valid {
		return "", ErrDriveNotFound
	}
	ext := strings.ToLower(filepath.Ext(n.Name))
	if _, ok := imageExts[ext]; !ok {
		return "", ErrDriveNotImage
	}

	srcAbs := s.BlobAbsPath(n.BlobPath.String)
	thumbsDir := filepath.Join(s.config.BasePath, "drive", "_thumbs")
	if err := os.MkdirAll(thumbsDir, 0755); err != nil {
		return "", err
	}
	thumbAbs := filepath.Join(thumbsDir, filepath.Base(n.BlobPath.String)+".jpg")

	if st, err := os.Stat(thumbAbs); err == nil && st.Size() > 0 {
		return thumbAbs, nil
	}

	mu := keyLock(thumbAbs)
	mu.Lock()
	defer mu.Unlock()
	// Double-check after acquiring the lock.
	if st, err := os.Stat(thumbAbs); err == nil && st.Size() > 0 {
		return thumbAbs, nil
	}

	img, err := imaging.Open(srcAbs, imaging.AutoOrientation(true))
	if err != nil {
		return "", fmt.Errorf("decode %s: %w", srcAbs, err)
	}

	bounds := img.Bounds()
	w, h := bounds.Dx(), bounds.Dy()
	if w == 0 || h == 0 {
		return "", fmt.Errorf("zero-dimension image")
	}
	// Don't upscale: if already small enough, just re-encode as JPEG.
	tw := thumbWidth
	if w < tw {
		tw = w
	}
	th := h * tw / w
	thumb := imaging.Resize(img, tw, th, imaging.Lanczos)

	tmp := thumbAbs + ".part"
	f, err := os.Create(tmp)
	if err != nil {
		return "", err
	}
	if err := jpeg.Encode(f, thumb, &jpeg.Options{Quality: 82}); err != nil {
		f.Close()
		_ = os.Remove(tmp)
		return "", err
	}
	if err := f.Close(); err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	if err := os.Rename(tmp, thumbAbs); err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	return thumbAbs, nil
}

// PurgeThumb removes the cached thumbnail for a blob (best-effort). Called
// when the underlying file is overwritten/deleted.
func (s *DriveService) PurgeThumb(blobPath string) {
	if blobPath == "" {
		return
	}
	thumbAbs := filepath.Join(s.config.BasePath, "drive", "_thumbs",
		filepath.Base(blobPath)+".jpg")
	_ = os.Remove(thumbAbs)
}
